lib/spack/spack/build_environment.py

```diff
@@ -835,154 +831,256 @@ def setup_package(pkg, dirty, context="build"):
-    environment variables), so this applies changes in a fixed order:
+        # Mark the starting point
+        if edge.parent is None:
+            modes[key] = Mode.ROOT
+            continue
 
-    - All modifications (custom and default) from external deps first
-    - All modifications from non-external deps afterwards
+        mode = modes[edge.parent]
 
-    With that order, `PrependPath` actions from non-external default
-    environment modifications will take precedence over custom modifications
-    from external packages.
+        # Nothing to propagate.
+        if not mode:
+            continue
 
-    A secondary constraint is that custom and default modifications are
-    grouped on a per-package basis: combined with the post-order traversal this
-    means that default modifications of dependents can override custom
-    modifications of dependencies (again, this would only occur for PATH,
-    CMAKE_PREFIX_PATH, or PKG_CONFIG_PATH).
+        # Dependending on the context, include particular deps from the root.
+        if Mode.ROOT in mode:
+            if context == "build":
+                if "build" in edge.deptypes or "test" in edge.deptypes:
+                    modes[key] |= Mode.BUILDTIME_DIRECT
+                if "link" in edge.deptypes:
+                    modes[key] |= Mode.BUILDTIME
+
+            elif context == "test":
+                if "run" in edge.deptypes or "test" in edge.deptypes:
+                    modes[key] |= Mode.RUNTIME_EXECUTABLE
+                elif "link" in edge.deptypes:
+                    modes[key] |= Mode.RUNTIME
+
+            elif context == "run":
+                if "run" in edge.deptypes:
+                    modes[key] |= Mode.RUNTIME_EXECUTABLE
+                elif "link" in edge.deptypes:
+                    modes[key] |= Mode.RUNTIME
+
+        # Propagate RUNTIME and RUNTIME_EXECUTABLE through link and run deps.
+        if (Mode.RUNTIME | Mode.RUNTIME_EXECUTABLE | Mode.BUILDTIME_DIRECT) & mode:
+            if "link" in edge.deptypes:
+                modes[key] |= Mode.RUNTIME
+            if "run" in edge.deptypes:
+                modes[key] |= Mode.RUNTIME_EXECUTABLE
+
+        # Propagate BUILDTIME through link deps.
+        if Mode.BUILDTIME in mode:
+            if "link" in edge.deptypes:
+                modes[key] |= Mode.BUILDTIME
+
+        # Finalize the spec; the invariant is that all in-edges are processed
+        # before out-edges, meaning that edge.parent is done.
+        if Mode.ADDED not in mode:
+            modes[edge.parent] |= Mode.ADDED
+            nodes_with_type.append((edge.parent, mode))
+
+    # Attach the leaf nodes, since we only added nodes
+    # with out-edges. Can this be improved?
+    for spec, mode in modes.items():
+        if mode and Mode.ADDED not in mode:
+            nodes_with_type.append((spec, mode))
+
+    return nodes_with_type
+
+
+def modifications_from_dag(
+    specs: Union[spack.spec.Spec, List[spack.spec.Spec]],
```

Comment:
Same as above, we can use:
```suggestion
    *specs: spack.spec.Spec,
```